### PR TITLE
Wildcards nodes and units

### LINKED ISSUE #185: Want ability to monitor using wildcards
It would be good if the monitors could support wildcards like "all services on node", or "on  all nodes".

### PR DESCRIPTION
**Pull request by engelmi**
This PR attempts to fix #185 by adding wildcard `*` support for nodes and units. 

In addition, it also contains:   
- Change to the `Subscribe` and `Unsubscribe` API to accept a list of node-unit-tuples
- Extending `hirtectl` by a (basic) implementation for creating a monitor + subscription(s)

### CODE REVIEW

File: src/agent/agent.c [hunk 1]
```diff
@@ -1076,6 +1087,11 @@ static int agent_method_unsubscribe(sd_bus_message *m, void *userdata, UNUSED sd
                 return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
         }
 
+        if (streq(unit, "*")) {
+                agent->wildcard_subscription_active = false;
```
Comment (alexlarsson) [inline src/agent/agent.c:1100]: Maybe check that wildcard_subscription_active is already set and error if not?

File: src/manager/monitor.c [hunk 2]
```diff
@@ -194,6 +195,84 @@ static int monitor_method_close(sd_bus_message *m, void *userdata, UNUSED sd_bus
...
+        bool success = true;
+
+        Node *current_node = NULL;
+        Node *next_node = NULL;
+        LIST_FOREACH_SAFE(nodes, current_node, next_node, manager->nodes) {
+                if (!subscribe_on_node(manager, monitor, current_node->name, unit)) {
+                        success = false;
+                        break;
+                }
+        }
```
Comment (alexlarsson) [inline src/manager/monitor.c]: There was already support in manager.c to handle node == NULL in subscribe meaning to match for all nodes. However, you removed that and replaced it with this loop instead. I don't really think this approach is better. You will just get more Subscription objects, and you need the cleanup unsubscribe below that wasn't otherwise needed.

Is there a pro to having one Subscription per node in this case?

File: src/manager/monitor.c [hunk 3]
```diff
@@ -194,6 +195,84 @@ static int monitor_method_close(sd_bus_message *m, void *userdata, UNUSED sd_bus
...
  ************************************************************************/
 
+static bool subscribe_on_node(Manager *manager, Monitor *monitor, const char *node, const char *unit) {
+        _cleanup_free_ char *unit_copy = NULL;
+        if (!copy_str(&unit_copy, unit)) {
+                return false;
+        }
+
+        char *saveptr = NULL;
+        char *unit_name = strtok_r(unit_copy, ",", &saveptr);
```
Comment (alexlarsson) [inline src/manager/monitor.c]: I don't like using stringly typed things like this. If we want to pass multiple units, we should do it with actually typed dbus arrays.

**Comment by alexlarsson** [comment]
I don't like how the unsubscribe API ends up with this new multi-unit/node Subscribe API. You can subscribe to some set, and then unsubscribe to a partial set of those, ending up with some weird in-between state.

I would prefer an API where subscribe added a set of subscriptions, and then returned some kind of identified for that set (the Subscribe object). Then you would pass that identifier to unsubscribe, and that particular Subscribe object is removed. You should not be able to construct subscription states other than those of specified by a whole subscribe call.

File: src/manager/monitor.c [hunk 4]
```diff
@@ -244,6 +248,45 @@ static int monitor_method_subscribe(sd_bus_message *m, void *userdata, UNUSED sd
...
+                                        m,
+                                        SD_BUS_ERROR_INVALID_SIGNATURE,
+                                        "Invalid subscription signature: %s",
+                                        strerror(-r));
+                }
+
+                if (is_wildcard(node)) {
+                        subscribe_on_all(manager, monitor, unit);
+                } else {
+                        subscribe_on_node(manager, monitor, node, unit);
```
Comment (alexlarsson) [inline src/manager/monitor.c]: I really would like a single Subscription objects with a list of units in it rather than multiple Subscriptions like this. Then Unsubscribe should just remove this one object.

File: src/manager/monitor.c (hunk 3 above)
Comment (engelmi) [reply src/manager/monitor.c]: I totally agree. This has been removed in one of the subsequent commits. By the latest rebase I cleaned up that commit as well - its now in the `SubscribeList` API method.

File: src/manager/monitor.c (hunk 2 above)
Comment (engelmi) [reply src/manager/monitor.c]: Not really. I wanted to move that logic into the monitor, but - thinking about it - it doesn't really fit there. This part has changed with the latest changes based on our discussion and is now similar to the previous implementation.

File: src/manager/monitor.c [hunk 5]
```diff
@@ -46,6 +46,38 @@ Subscription *create_monitor_subscription(Monitor *monitor, const char *node, co
...
+
+        LIST_APPEND(units, sub->subscribed_units, su);
+        return true;
+}
+
+bool subscription_has_node_wildcard(Subscription *sub) {
+        return sub->node != NULL && streq(sub->node, SYMBOL_WILDCARD);
+}
+
+bool subscription_has_unit_wildcard(Subscription *sub) {
```
Comment (alexlarsson) [inline src/manager/monitor.c]: I would probably do the is_wildcard() check in add_unit() and just set an is_wildcard boolean which you return from here.

File: src/manager/node.c [hunk 6]
```diff
@@ -174,6 +191,14 @@ void node_unref(Node *node) {
                 proxy_dependency_free(dep);
         }
 
+        Subscription *sub = NULL;
+        Subscription *next_sub = NULL;
+        LIST_FOREACH_SAFE(wildcard_subscriptions, sub, next_sub, node->wildcard_subscriptions) {
+                manager_remove_subscription(node->manager, sub);
+                LIST_REMOVE(wildcard_subscriptions, node->wildcard_subscriptions, sub);
+                subscription_unref(sub);
```
Comment (alexlarsson) [inline src/manager/node.c]: This part doesn't look right to me. Yes, we should maybe remove the sub from wildcard_subscriptions, because node_subscribe_wildcard_unit() added it there. However, we should not unref it, because we didn't ref it.  Alterntively we should ref the sub when adding it, but I don't think that is right, as we risk creating cycles. And anyway the node will not be unref:ed until shutdown, so lifetimes are not really a problem.

And we shouldn't call manager_remove_subscription(), which will remove the sub from other nodes too. That should only happen when the subscription is unsubscribed, which comes from the monitor.

File: src/manager/node.c [hunk 7]
```diff
@@ -799,31 +884,33 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
...
-                        }
-                }
 
-                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
-                        Subscription *sub = usub->sub;
-
-                        int r = sub->handle_unit_removed(sub->monitor, node->name, sub->unit, "virtual");
-                        if (r < 0) {
-                                hirte_log_error("Failed to emit UnitRemoved signal");
+                                if (send_state_change) {
```
Comment (alexlarsson) [inline src/manager/node.c]: This changes the order of the unit removed and the state changed events.
Also, I think i had some reentrency issue in  the case where handle_unit_removed destroyed the monitor, which made me do the double LIST_FOREACH_SAFE loops that you merged here.

File: src/manager/node.c [hunk 8]
```diff
@@ -799,31 +884,33 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
                 }
 
                 usubs->loaded = false;
+
                 UnitSubscription *usub = NULL;
```
Comment (alexlarsson) [inline src/manager/node.c]: Hmm, this only sends the virtual events for non-wildcard subscriptions. Also, we don't even track the last state for such subscriptions, so we can't sent any state change event...

File: src/manager/node.c [hunk 9]
```diff
@@ -1397,105 +1484,154 @@ static void node_send_agent_subscribe_all(Node *node) {
...
-                                        sub->unit,
-                                        active_state_to_string(usubs->active_state),
-                                        usubs->substate ? usubs->substate : "invalid",
-                                        "virtual");
+                LIST_APPEND(subs, usubs->subs, steal_pointer(&usub));
+
+                /* We know this is loaded, so we won't get notified from
+                   the agent, instead send a virtual event here. */
+                if (usubs->loaded) {
+                        int r = sub->handle_unit_new(sub->monitor, node->name, sub_unit->name, "virtual");
```
Comment (alexlarsson) [inline src/manager/node.c]: Same here, none of these virtual events are sent for a wildcard subscription.

File: src/manager/node.h [hunk 10]
```diff
@@ -50,6 +50,7 @@ struct Node {
         LIST_HEAD(ProxyMonitor, proxy_monitors);
         LIST_HEAD(ProxyDependency, proxy_dependencies);
 
+        LIST_HEAD(Subscription, wildcard_subscriptions);
```
Comment (alexlarsson) [inline src/manager/node.h]: Alternatively you could store the wildcard subscription in the unit_subscription hashmap under the "*" key. Not sure which is cleaner.

File: src/manager/node.c [hunk 11]
```diff
@@ -1397,105 +1484,154 @@ static void node_send_agent_subscribe_all(Node *node) {
...
 
-        LIST_FOREACH(subs, usub, usubs->subs) {
-                if (usub->sub == sub) {
-                        found = usub;
-                        break;
-                }
+void node_unsubscribe_wildcard_unit(Node *node, Subscription *sub) {
+        LIST_REMOVE(wildcard_subscriptions, node->wildcard_subscriptions, sub);
+        if (!node_has_wildcard_subscriptions(node)) {
+                /* First wildcard sub to this node, pass to agent */
```
Comment (alexlarsson) [inline src/manager/node.c]: Should comment be "Last wildcard" maybe?

File: src/manager/monitor.c (hunk 5 above)
Comment (engelmi) [reply src/manager/monitor.c]: Good point, but I actually was able to remove it due to merging the wildcard list into the existing unit_subscription hashmap.

File: src/manager/node.c (hunk 6 above)
Comment (engelmi) [reply src/manager/node.c]: Removed this part completely since I merged the `wildcard_subscriptions` into the existing hashmap for the `unit_subscriptions`.

File: src/manager/node.c (hunk 7 above)
Comment (engelmi) [reply src/manager/node.c]: Ups, you are right - I didn't think about the order here. Reverted it.

File: src/manager/node.c (hunk 11 above)
Comment (engelmi) [reply src/manager/node.c]: Yes, copy & paste error :sweat_smile:

File: src/manager/node.h (hunk 10 above)
Comment (engelmi) [reply src/manager/node.h]: I tried that approach and I could simplify a bit of code - so I'd go with that.

File: src/manager/node.c (hunk 8 above)
Comment (engelmi) [reply src/manager/node.c]: As we discussed, I changed this so that virtual `UnitNew` and `UnitRemoved` (also StateChanged) signals are emitted. This way the monitor gets notified if, for example, the node disconnects and reconnects, and can act upon it (e.g. query current state of special units).

File: src/agent/agent.c (hunk 1 above)
Comment (engelmi) [reply src/agent/agent.c:1100]: Yes, added that check.

**Comment by rhatdan** [comment]
@engelmi needs a rebase.

**Comment by engelmi** [comment]
> @engelmi needs a rebase.

Thanks for the notification. Done :heavy_check_mark:

File: doc/dbus-interfaces.md [hunk 12]
```diff
@@ -76,16 +76,24 @@ Object path: `/org/containers/hirte/monitor/$id`
...
-  * `Subscribe(in node s, in unit s)`
+  * `Subscribe(in node s, in unit s, out id u)`
 
-    Subscribe to changes in properties of a given unit on a given node, or all nodes if the node name is empty. This
-    will emit the signal `UnitChanged` for all matching units in the system, and then again whenever one of the
-    properties of the unit changes.
+    Subscribe to changes in properties of a given unit on a given node. Passing a wildcard `*` to either `node` or
+    `unit` will subscribe to all nodes or all units on the given node. This will emit the signal `UnitChanged` for all
+    matching units in the system, and then again whenever one of the properties of the unit changes. Returns an
+    identifier `id` used for a subsequent `Unsubscribe`.
```
Comment (alexlarsson) [inline doc/dbus-interfaces.md:84]: This isn't needed in this PR, but at some point we need to extend this with a fuller description of the events, including the virtual events (and how they differ for wildcarded matches).

File: src/agent/agent.c [hunk 13]
```diff
@@ -1043,6 +1044,19 @@ static int agent_method_subscribe(sd_bus_message *m, void *userdata, UNUSED sd_b
...
         }
 
+        if (is_wildcard(unit)) {
+                if (agent->wildcard_subscription_active) {
+                        return sd_bus_reply_method_errorf(
+                                        m, SD_BUS_ERROR_FAILED, "Already wildcard subscribed");
+                }
+                agent->wildcard_subscription_active = true;
+
+                AgentUnitInfo info = { NULL, (char *) unit, true, true, UNIT_ACTIVE, NULL };
```
Comment (alexlarsson) [inline src/agent/agent.c]: UNIT_ACTIVE for a "*" unit seems a bit weird to me, but unfortunately I don't really see a better alternative, and I don't think we want to add new possible states.

Comment (alexlarsson) [reply src/agent/agent.c]: I think we should have a substate though.

File: src/manager/node.c [hunk 14]
```diff
@@ -260,53 +276,97 @@ static int node_match_unit_properties_changed(sd_bus_message *m, void *userdata,
...
-        usubs = hashmap_get(node->unit_subscriptions, &key);
-        if (usubs == NULL) {
-                return 0;
+        UnitSubscriptions *usubs = hashmap_get(node->unit_subscriptions, &key);
+        if (usubs != NULL) {
+                usubs->loaded = true;
+                if (is_wildcard(unit)) {
+                        usubs->active_state = UNIT_ACTIVE;
+                        free(usubs->substate);
+                        usubs->substate = strdup("running");
```
Comment (alexlarsson) [inline src/manager/node.c:364]: This initializes the virtual substate for a wildcard to "running", so maybe we should use that in the agent too?

File: src/manager/node.c [hunk 15]
```diff
@@ -260,53 +276,97 @@ static int node_match_unit_properties_changed(sd_bus_message *m, void *userdata,
...
+        const UnitSubscriptionsKey wildcard_key = { (char *) SYMBOL_WILDCARD };
+        UnitSubscriptions *usubs_wildcard = hashmap_get(node->unit_subscriptions, &wildcard_key);
+        if (usubs_wildcard != NULL) {
+                UnitSubscription *usub = NULL;
+                UnitSubscription *next_usub = NULL;
+                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs_wildcard->subs) {
+                        Subscription *sub = usub->sub;
+
+                        /* only emit signal once per monitor */
+                        if (usubs != NULL &&
```
Comment (alexlarsson) [inline src/manager/node.c]: This is called in several places, so maybe you can move the usubs != NULL check into unit_subscription_contains_monitor.

File: src/manager/node.c [hunk 16]
```diff
@@ -260,53 +276,97 @@ static int node_match_unit_properties_changed(sd_bus_message *m, void *userdata,
...
-        usubs = hashmap_get(node->unit_subscriptions, &key);
-        if (usubs == NULL) {
-                return 0;
+        UnitSubscriptions *usubs = hashmap_get(node->unit_subscriptions, &key);
+        if (usubs != NULL) {
+                UnitSubscription *usub = NULL;
+                UnitSubscription *next_usub = NULL;
+                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
+                        Subscription *sub = usub->sub;
+                        int r = sub->handle_unit_property_changed(sub->monitor, node->name, unit, interface, m);
```
Comment (alexlarsson) [inline src/manager/node.c]: This is not your fault, but I think there is another issue here. What if a single monitor has multiple subscriptions that overlap. We will then send multiple events to this single monitor.

I wonder if we should have the two loops just compute a vector of of (unique) monitors, and then loop over those and call handle_unit_property_changed.

File: src/manager/node.c [hunk 17]
```diff
@@ -799,31 +897,45 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
...
-                                                sub->monitor,
-                                                node->name,
-                                                sub->unit,
-                                                active_state_to_string(usubs->active_state),
-                                                usubs->substate,
-                                                "virtual");
-                                if (r < 0) {
-                                        hirte_log_error("Failed to emit UnitRemoved signal");
+
+                                SubscribedUnit *unit_sub = NULL;
```
Comment (alexlarsson) [inline src/manager/node.c]: This seems wrong to me. On subscribe we take the list of subscribed units in the user-specified subscription and for each of the specified units create a new UnitSuibscription and put in the hashtable. So, now that you interate over all the hashtable entries you should get every unit name ever requested. So, there is no need to now iterate over all the units in this subscription again.

In fact, it is very wrong. The send_state_change and the active_state computed above that we are sending *only* applies to the unit we're currently looking at from the hashtable.

File: src/manager/node.c [hunk 18]
```diff
@@ -799,31 +897,45 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
...
+                next_usub = NULL;
                 LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
                         Subscription *sub = usub->sub;
 
-                        int r = sub->handle_unit_removed(sub->monitor, node->name, sub->unit, "virtual");
-                        if (r < 0) {
-                                hirte_log_error("Failed to emit UnitRemoved signal");
+                        SubscribedUnit *unit_sub = NULL;
+                        SubscribedUnit *next_unit_sub = NULL;
+                        LIST_FOREACH_SAFE(units, unit_sub, next_unit_sub, sub->subscribed_units) {
```
Comment (alexlarsson) [inline src/manager/node.c]: same here

File: src/manager/node.c [hunk 19]
```diff
@@ -1394,104 +1506,112 @@ static void node_send_agent_subscribe_all(Node *node) {
...
-                                        sub->unit,
-                                        active_state_to_string(usubs->active_state),
-                                        usubs->substate ? usubs->substate : "invalid",
-                                        "virtual");
+                LIST_APPEND(subs, usubs->subs, steal_pointer(&usub));
+
+                /* We know this is loaded, so we won't get notified from
+                   the agent, instead send a virtual event here. */
+                if (usubs->loaded) {
+                        int r = sub->handle_unit_new(sub->monitor, node->name, sub_unit->name, "virtual");
```
Comment (alexlarsson) [inline src/manager/node.c]: This is also potentially sending the same "new" unit twice if you subscribe to the same unit twice. I wonder if we should track that we sent it? This is kinda fringe semantics though.

File: src/agent/agent.c (hunk 13 above)
Comment (engelmi) [reply src/agent/agent.c]: Yes, I agree that the `UnitNew` doesn't really fit. The unit new signal only contains the name of the unit and the reason - so every other information like the active state is not used. `AgentInfo` is only created for the sake of calling the `agent_emit_unit_new` function. Maybe replacing `UNIT_ACTIVE` with a -1 is more clear?

File: src/manager/node.c (hunk 14 above)
Comment (engelmi) [reply src/manager/node.c:364]: Contrary to the manager, in the agent the wildcard is only represented as a boolean flag and is not traced in any other way. If active, the agent checks both, the subscription of the specific unit (just like before) and, in addition, the wildcard flag.

File: src/manager/node.c (hunk 17 above)
Comment (engelmi) [reply src/manager/node.c]: Good catch! I remove the unnecessary loop.

File: doc/dbus-interfaces.md (hunk 12 above)
Comment (engelmi) [reply doc/dbus-interfaces.md:84]: Good point, I created [an issue](https://github.com/containers/hirte/issues/261) to keep track of this task.

File: src/manager/node.c (hunk 16 above)
Comment (engelmi) [reply src/manager/node.c]: I moved the `handle_xyz` functions from the `Subscription` to the `Monitor`, which seems to be a better fit. When I tried to compute a unique list of monitors with only a hashmap (using the monitor id as key), that resulted in segfaults due to the `handle_xyz` functions being NULL - since the hashmap copies everything. I decided to use a hashmap just for the ids to track which monitor is already added to a list of unique monitors. 
This also results in the `unit_subscription_contains_monitor` to be removed.

File: src/manager/node.c (hunk 15 above)
Comment (engelmi) [reply src/manager/node.c]: Done. But the `unit_subscription_contains_monitor` has been removed later in favor of building a unique monitor list.

File: src/agent/agent.c (hunk 13 above)
Comment (alexlarsson) [reply src/agent/agent.c]: _UNIT_ACTIVE_STATE_INVALID maybe? Although I don't really think it matters. Lets leave it.

File: src/manager/node.c [hunk 20]
```diff
@@ -262,18 +246,38 @@ static int node_match_job_state_changed(
...
+        struct hashmap *unique_monitor_ids = hashmap_new(
+                        sizeof(UnitSubscriptions),
+                        0,
+                        0,
+                        0,
+                        monitor_hashmap_hash,
+                        monitor_hashmap_compare,
+                        NULL,
+                        NULL);
+        if (unique_monitor_ids == NULL) {
```
Comment (alexlarsson) [inline src/manager/node.c]: This use of hash is rather weird. What you want to do is put the actual monitor pointers into the hash, and then just return the resulting hashtable and have the caller use hashmap_iter() to loop over the monitors.

You need:

```
static uint64_t monitor_hashmap_hash(const void *item, uint64_t seed0, uint64_t seed1) {
   return (uint64_t)item;
}

static int monitor_hashmap_compare(const void *a, const void *b, UNUSED void *udata) {
   return a == b;
}
```

Then pass size as sizeof (void *), and on insert:

```
hashmap_set(unique_monitors, mo); /* Ignore result, no need to free old value */
if (hashmap_oom(unique_monitor_ids)) {
 ...
}
```

File: src/manager/node.c [hunk 21]
```diff
@@ -334,38 +380,18 @@ static int node_match_unit_new(sd_bus_message *m, void *userdata, UNUSED sd_bus_
...
-                                        hirte_log_errorf("Failed to emit UnitNew signal: %s", strerror(-r));
-                                }
-                        }
+        Monitor *unique_monitor = NULL;
+        Monitor *unique_monitor_next = NULL;
+        LIST_FOREACH_SAFE(
+                        monitors,
+                        unique_monitor,
+                        unique_monitor_next,
+                        node_compute_unique_monitor_list(node, unit)) {
```
Comment (alexlarsson) [inline src/manager/node.c]: Is this really right if `unit == "*"`? I'm not sure. Seems like we may sometimes send both virual new "*" from the manager and then the agent in some cases. Needs more careful analysis.

File: src/manager/node.c [hunk 22]
```diff
@@ -901,38 +891,27 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
 
                 if (send_state_change) {
                         LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
```
Comment (alexlarsson) [inline src/manager/node.c]: I wonder if this should not also call node_compute_unique_monitor_list() for name=usubs->name? I can imagine multiple subscriptions for the same monitor?

I guess that would behave somewhat different, as it would also send events on disconnect for actual units to wildcarded monitors. Maybe node_compute_unique_monitor_list() can have a bool include_wildcards argument?

File: src/manager/monitor.c [hunk 23]
```diff
@@ -37,11 +37,6 @@ Subscription *create_monitor_subscription(Monitor *monitor, const char *node) {
         subscription->monitor = monitor_ref(monitor);
         subscription->free_monitor = (free_func_t) monitor_unref;
 
-        subscription->handle_unit_new = monitor_on_unit_new;
```
Comment (alexlarsson) [inline src/manager/monitor.c:41]: Ideally you should break out this change to a separate commit.

File: src/manager/node.c (hunk 20 above)
Comment (alexlarsson) [reply src/manager/node.c]: Ah, actually, compare should be if (a == b) return 0; else return 1;

File: src/manager/node.c (hunk 21 above)
Comment (engelmi) [reply src/manager/node.c]: I think this should be alright for `*`.  The agent sends only a virtual UnitNew for wildcards when such a subscription is made - which is forwarded by the manager to the monitor. Since wildcard units and "regular" units are stored in the `unit_subscriptions` of the respective node, `node_compute_unique_monitor_list` prevents duplicate emitting the UnitNew signal to the monitor. 
If the node disconnects, then the monitor with the wildcard gets treated the same as regular unit subs and a StateChanged (with the `agent-offline`) substate as well as a UnitRemoved gets emitted for it. On a reconnect of the agent, the manager subscribes with the wildcard again and we receive another UnitNew signal. 
Here is a sample output for subscribing, disconnecting the agent and reconnecting again:
```
Monitor path: /org/containers/hirte/monitor/5
Subscribing to node 'laptop' and unit '*'
[laptop] *
        Unit created (reason: virtual)
[laptop] *
        Unit state changed (reason: virtual)
        Active: inactive (agent-offline)
[laptop] *
        Unit removed (reason: virtual)
[laptop] *
        Unit created (reason: virtual)

```

File: src/manager/node.c (hunk 22 above)
Comment (engelmi) [reply src/manager/node.c]: Yes, good point. I'll change it to use the `node_compute_unique_monitor_list` instead. 
I think we want to have the same behavior for wildcard subs as we have for regular ones. This ensures that the user of the monitor knows when an agent disconnected and it might want to catch up with the latest state if it reconnects.

File: src/manager/monitor.c (hunk 23 above)
Comment (engelmi) [reply src/manager/monitor.c:41]: Done.

**Comment by alexlarsson** [comment]
After some testing of monitors when shutting down hirte i got:
```
hirte: ../src/manager/node.c:97: unit_subscriptions_clear: Assertion `LIST_IS_EMPTY(usubs->subs)' failed.
```

**Comment by alexlarsson** [comment]
Reproducible by ctrl-c hirte when a "*" unit monitor is active.

File: src/manager/node.c [hunk 24]
```diff
@@ -791,31 +870,39 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
...
+                                                        usubs->substate,
+                                                        "virtual");
+                                        if (r < 0) {
+                                                hirte_log_error("Failed to emit UnitStateChanged signal");
+                                        }
                                 }
                         }
-                }
 
-                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
```
Comment (alexlarsson) [inline src/manager/node.c:821]: I think you need to call node_compute_unique_monitor_list() again here, because the call to handle_unit_state_changed() above may have removed the monitor.

**Comment by engelmi** [comment]
> After some testing of monitors when shutting down hirte i got:
> 
> ```
> hirte: ../src/manager/node.c:97: unit_subscriptions_clear: Assertion `LIST_IS_EMPTY(usubs->subs)' failed.
> ```

Yes, this issue is also occurring on the main branch. I'll try to investigate and fix it, but if its more work, I'd rather pull this out of this PR and create a small gh issue (which I'd tackle afterwards) since this PR has been open for quite a while. What do you think? @alexlarsson

**Comment by alexlarsson** [comment]
@engelmi I think the two commits i pushed should fix that.

**Review by alexlarsson — APPROVED** [review]
LGTM
